sdks/python/apache_beam/transforms/core.py

```diff
@@ -1970,10 +1985,14 @@ def add_input_types(transform):
       return combined
 
     if self.has_defaults:
-      combine_fn = (
-          self.fn if isinstance(self.fn, CombineFn) else
-          CombineFn.from_callable(self.fn))
-      default_value = combine_fn.apply([], *self.args, **self.kwargs)
+      combine_fn = copy.copy(
```

Comment:
What is the reason for a shallow copy here?

Reply:
Better protection against potential side effects.

If using default values, `CombineFn.apply` is called at pipeline construction time. `CombineFn.setup` and `CombineFn.teardown` are called along with it. The same instance of CombineFn is then serialized and sent to runner. I think it would be better to perform initial `CombineFn.apply` on a copy, so that the state of the instance is not polluted.

Reply:
I see. I wonder if calling setup/teardown on during pipeline submission may be undesireable in some cases.
One option to offer flexibility is to use introduce default_value() method in CombineFn, and move the setup/teardown call in that method, trying this out in: https://github.com/apache/beam/pull/13081.

Reply:
Left some comments over there